Ability to delete small volume elements

GiudGiud opened the pull request:
There's a good cleanup of MeshRepairGenerator included.

GiudGiud committed:
Add option to delete 0 or small absolute volume elements

GiudGiud committed:
Split off node overlap code

GiudGiud committed:
Add an option to consider the volume instead of the absolute volume
I m thinking we re just going to delete negative volume elements and get on our way

GiudGiud committed:
Touch ups

GiudGiud committed:
Rework small element deletion to avoid merging the nodes into the center
It s a bad idea, there's no way to recover from that it breaks all the neighbor elements

Break off subdomain element-type-based separation into its own function

Dont merge multiple nodes in the same element, not a good idea either

GiudGiud committed:
Split off small-volume deletion into a parsed element deletion generator
- more caability which is nice nice
- all the logic in ElementDeletionGeneratorBase comes in, which is big time nice nice

GiudGiud committed:
Add doco for the new MG

GiudGiud committed:
Add tests for parsed element deletion generator
refs #25397

GiudGiud commented:
oh CSV diff could be my fault from changing the transfer to the non-deprecated one..

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1748220/) on ec849cc wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/25399/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1748263/) on ec849cc wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>395233</th>
      <th colspan=4>#25399 ec849c</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.66%</td>
      <td>85.66%</td>
      <td>-0.01%</td>
      <td>89.25%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>90538</td>
      <td>90566</td>
      <td>+28</td>
      <td>83</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15154</td>
      <td>15165</td>
      <td>+11</td>
      <td>10</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25399/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25399/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/25399/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/25399/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/25399/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/25399/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/25399/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/25399/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/25399/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/25399/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/25399/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/25399/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/25399/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/25399/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/25399/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/25399/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/25399/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/25399/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/25399/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/25399/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/25399/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/25399/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/25399/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/25399/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/25399/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/25399/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/25399/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/25399/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/25399/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/25399/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 89.25% is less than the suggested 90.0%



This comment will be updated on new commits.

roystgnr commented on framework/src/meshgenerators/MeshRepairGenerator.C:
```diff
@@ -61,112 +63,136 @@ MeshRepairGenerator::generate()
+    separateSubdomainsByElementType(mesh);
+
+  mesh->set_isnt_prepared();
+  return dynamic_pointer_cast<MeshBase>(mesh);
```
Isn't `return mesh;` enough here?

GiudGiud replied:
I think it s tradition to do that

GiudGiud replied:
I ll change

roystgnr commented on framework/src/meshgenerators/MeshRepairGenerator.C:
```diff
@@ -61,112 +63,136 @@ MeshRepairGenerator::generate()
-            if (MooseUtils::absoluteFuzzyEqual(x_node, x_elem_node, tol) &&
-                MooseUtils::absoluteFuzzyEqual(y_node, y_elem_node, tol) &&
-                MooseUtils::absoluteFuzzyEqual(z_node, z_elem_node, tol))
+            // Merging two nodes from the same element is almost never a good idea
```
Good catch.  You'd think nobody would ever do this on purpose, but it's apparently quite popular for some codes to simulate a prism by collapsing one Hex face into an edge and then adding some constraints to prevent a discontinuity there.

roystgnr commented on framework/src/meshgenerators/MeshRepairGenerator.C:
```diff
@@ -61,112 +63,136 @@ MeshRepairGenerator::generate()
-        }
+        auto new_id = next_block_id + i++;
+        // Create blocks when a block has multiple element types
+        mesh->subdomain_name(new_id) = mesh->subdomain_name(id) + "_" + Moose::stringify(type);
```
This is really nice.  IMHO we really ought to be doing this at the libMesh level, eventually.  The conflict between "ExodusII blocks can't be heterogeneous" and "ExodusII block ids are easy to lazily map 1-to-1 with libMesh subdomain ids" should *not* have resolved out to "libMesh subdomains can't be heterogeneous", but rather to "libMesh::ExodusII_IO should be less lazy".

GiudGiud replied:
yeah that could also be the fix. I think there is some value in splitting the subdomains even outside of exodus IO lack of support. 

what if we introduce element-specific residual computation for example

roystgnr commented on framework/include/meshgenerators/ParsedElementDeletionGenerator.h:
```diff
@@ -0,0 +1,38 @@
+ * Deletes elements based on the evaluation of a parsed expression, involving the coordinates of
+ * their vertex average and their volume
+ */
+class ParsedElementDeletionGenerator : public ElementDeletionGeneratorBase,
```
It's an existing `ElementDeletionGeneratorBase` issue, nothing that needs to be fixed in this PR, but:

If we delete a child element of a libMesh mesh without also deleting its parent, or vice-versa, the resulting mesh is invalid, almost certain to fail assertions in `devel` or segfault in `opt`.  And choosing based on element volume, where parents will exceed their children by `2^d`, seems especially likely to trigger that.  This scares me.

I'm having trouble picturing the use case for this too.  Do we want to remove part of our domain?  Generally an efficient FEM analysis should use smaller elements in the most important parts of the domain; why are smaller elements in particular the parts we want to instead remove entirely?

GiudGiud replied:
I ll take a look separately at this child/parent deletion.

I have a bad mesh with tiny and bad elements on the boundaries. Deleting them lets the simulation go forward but does not inflence results afaicansee

At the end of the day the mesh should be redone, it s not my decision to make

roystgnr replied:
Would it make sense to select for deletion based on something other than `Elem::volume()`?  `Elem::quality(ASPECT_RATIO)` or another quality measure?  There's no real a-priori way to define "bad elements" (e.g. flow simulations often grade sharply into boundaries, so the final layer of elements is tiny *and* high aspect-ratio) but in general bad shape tends to screw up conditioning in a less preconditioner-recoverable way than bad size does.

Would it make sense to filter ... somehow ... for elements that are on the boundary?  It sounds like the answer to "Do we want to remove part of our domain?" is "Yes!", but it's a case where we can get away with removing boundary elements specifically because that just shrinks the domain a tiny bit; accidentally adding gaps mid-domain might be more serious.  The trouble is, I can't figure out how to do this properly; the obvious algorithm would require reimplementing half of the base class.

I guess this works as-is, as a "the mesh should be redone but it won't be" hack.

GiudGiud replied:
but actually following that same logic of "this is my need right now" I added the element id. With this, I should be able to finish torturing Kyle's mesh and say "if you do not redo your mesh or pay for fancier libmesh smoothing, this is as good as I can get that mesh to be"

roystgnr reviewed (dismissed):
Approving, because I can't deny that the new class does what it says on the label, but I'd sure feel better if I first understood why we'd ever want to do that.

(the other fixes and refactoring I'm happy with unconditionally)

lindsayad commented on framework/src/meshgenerators/MeshRepairGenerator.C:
```diff
@@ -61,112 +63,136 @@ MeshRepairGenerator::generate()
   }
+  _console << "Number of overlapping nodes which got merged: " << num_fixed_nodes << std::endl;
+  if (mesh->allow_renumbering())
+  {
+    mesh->remove_orphaned_nodes();
+    mesh->renumber_nodes_and_elements();
+  }
```
```suggestion
    mesh->renumber_nodes_and_elements();
```
Orphaned nodes are automatically removed in `renumber_nodes_and_elements`

lindsayad commented on framework/src/meshgenerators/MeshRepairGenerator.C:
```diff
@@ -61,112 +63,136 @@ MeshRepairGenerator::generate()
+  std::set<subdomain_id_type> ids;
+  mesh->subdomain_ids(ids);
+  // loop on sub-domain
+  for (auto & id : ids)
```
```suggestion
  for (const auto id : ids)
```

lindsayad commented on framework/src/meshgenerators/MeshRepairGenerator.C:
```diff
@@ -61,112 +63,136 @@ MeshRepairGenerator::generate()
-      // loop on elements within this sub-domain
-      for (auto & elem : mesh->active_subdomain_elements_ptr_range(id))
-        types.insert(elem->type());
-
-      if (types.size() > 1)
+      auto next_block_id = MooseMeshUtils::getNextFreeSubdomainID(*mesh);
```
`getNextFreeSubdomainID` will perform a global communication. However, `types` may not be globally consistent as the code is now and so it's possible that some processes will proceed to the global communication and others will not. Probably the thing to do is to communicate `types` before this `if` condition

GiudGiud replied:
changed

GiudGiud replied:
mesh is serialized btw, which for us currently means replicated.
Thanks for the idea though, this is future-proof now

lindsayad replied:
This would be a problem even with a serial mesh as you are determining `types` by looping over local elements

GiudGiud replied:
I m using mesh->active_subdomain_elements_ptr_range(id))
not mesh->active_subdomain_(local)_elements_ptr_range(id)) (which does not exist)

and if I m reading this right, it calls active_subdomain_elements_begin ...
instead of active_local_subdomain_elements_begin (this one does exist)

GiudGiud replied:
now is active_local_subdomain_elements_begin about local subdomains or local elements lol

no idea. Thank you for the fix

lindsayad replied:
> m using mesh->active_subdomain_elements_ptr_range(id))
> not mesh->active_subdomain_(local)_elements_ptr_range(id)) (which does not exist)

Ah I must have read it wrong. But glad we're in both a good current and future state

lindsayad commented on framework/src/meshgenerators/ParsedElementDeletionGenerator.C:
```diff
@@ -0,0 +1,63 @@
+ParsedElementDeletionGenerator::shouldDelete(const Elem * elem)
+{
+  for (const auto i : make_range(3))
+    _func_params[i] = elem->vertex_average()(i);
```
I would evaluate `vertex_average()` outside of the loop

GiudGiud replied:
done

lindsayad commented on test/tests/meshgenerators/parsed_element_deletion_generator/tests:
```diff
@@ -0,0 +1,21 @@
+    input = 'cut_in_half.i'
+    jsondiff = 'cut_in_half_json.json'
+    recover = false
+
```
```suggestion
```

lindsayad commented on test/tests/meshgenerators/parsed_element_deletion_generator/tests:
```diff
@@ -0,0 +1,21 @@
+    input = 'cut_the_small.i'
+    jsondiff = 'cut_the_small_json.json'
+    recover = false
+
```
```suggestion
```

GiudGiud committed:
Address review:
- more const-ness
- be more careful with a call that requires global communication

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

GiudGiud committed:
Add capability to delete elements by their id
